.github/workflows/backend-cd-gcp.yml

```diff
@@ -44,15 +44,26 @@ jobs:
 
           echo "DOCKER_IMAGE=$DOCKER_REGISTRY_HOST/$PROJECT_ID/titus/backend:$GITHUB_SHA" >> $GITHUB_ENV
 
-      - name: Build container
+      - name: Install Grype tool (vulnerability scan)
+        run: curl -sSfL https://raw.githubusercontent.com/anchore/grype/main/install.sh | sh -s -- -b /usr/local/bin
```

Comment:
is this the recommended way to install grype? asking because we would usually avoid running anything that's on master, we tend to prefer using something that's published so upgrading to a specific version is an intentional thing, rather than using "whatever is latest"

Reply:
Yes, that's its recommended way (https://github.com/anchore/grype#recommended). Grype checks for the existence of a new tool version and then tries to update its vulnerability db in each execution before doing anything else. It's in its nature to be dynamic and fits perfectly in an ephemeral environment like a Github workflow.

Anyway, it's possible to pin a release version, but I wouldn't recommended it. I think it's better to have the latest available version so we get the full capabilities of Grype in each workflow execution.

Reply:
I trust your advice on this but I should point out that not too long ago a tool for code coverage called codecov had a vulnerability for this specific reason. running code in the master branch of any repository is not really a very safe approach because it's "too close" to any potentially harmful activity of an attacker. one commit from a malicious user and we're screwed because we're picking it up right away. instead if we rely on a "released" version of the tool, there are most steps  that an attacker would have to take in order to get malicious code into user's code. if there is a way to use a specific version but loosely, as we're doing with Node 16, I would prefer it

Reply:
Grype's installation is not using the master branch HEAD but the latest release available at GitHub. If you're curious about this, have a look at the installation script and look for the `tag_to_version` function. That's why I rely on its recommended installation. I totally agree with you about the dangers of using an arbitrary commit to install a tool.

Reply:
that's a good thing but nonetheless the entry point is an executable script on the main branch. if an attacker manages to push a commit there we're screwed. see https://about.codecov.io/security-update/

Reply:
Do you mean the `install.sh` script? In that case, even pinning the release version wouldn't help as we'd be using the very same installation script. Therefore, to be "completely" safe we should build our own installation procedure (downloading the proper binary from GitHub) or go through the Docker option (but I haven't used it in Github workflows, so I don't know if that would work).

Reply:
One option that I would be more comfortable with is run it from docker instead. They publish images there, even though they don't have documentation. See https://github.com/anchore/grype/issues/388

Reply:
Yes, I'm aware they're quite lacking in that aspect.

Let me have a look at the Docker option and see what I can come up with.

Reply:
It's not a big deal really, but I must say I find it somewhat ironic that a tool for security scanning has such an obvious flaw. All it takes is a malicious commits to screw up users.

Reply:
I've tested the Docker image and, while it works great in my local environment from the CLI, it fails when run from a workflow (it doesn't seem to be compatible with GitHub's workflows requirements): https://github.com/nearform/titus/runs/4701694131?check_suite_focus=true

Using the Grype Docker container would also imply to fiddle with volumes to be able to access the Docker tar file and export the vulnerability report, as well as running certain actions inside a container that may not be ready for the task.

As building our own Grype image would take such an investment of effort and time, I consider downloading a specific binary a more profitable path to explore, which is what I'm starting now.

Reply:
that's ok, let's stick with the shell script then. it was an interesting learning opportunity anyway!

Reply:
I'm not using the installation script anymore (even if they have protected their `main branch). I'm working on issuing several sequential commands so a specific release archive is downloaded and extracted and the binary installed in `/usr/local/bin`.

Reply:
I've learned about using Docker containers in a GitHub workflow, though.

Reply:
Thinking about it, I don't need the installation script's extra features, like autodetecting the target platform (arch, OS, etc.). The workflow uses `ubuntu-latest` as runner, so we can go safely with the `linux-amd64` binary.